File: typescript/day12_part1_2023.ts
const fs = require('fs');

interface Row {
  springs: string;
  group: number[];
}

function parseInput(input: string[]): Row[] {
  return input.map(line => {
    const [springs, groupStr] = line.split(' ');
    const group = groupStr.split(',').map(Number);
    return { springs, group };
  });
}

function countArrangementsRecursive(row: Row, iSprings: number, iGroup: number, iContiguousDamaged: number, cache: { [key: string]: number }): number {
  if (iSprings === row.springs.length) {
    if (iGroup === row.group.length && iContiguousDamaged === 0) return 1;
    if (iGroup === row.group.length - 1 && iContiguousDamaged === row.group[iGroup]) return 1;
    return 0;
  }

  const cacheKey = `${iSprings},${iGroup},${iContiguousDamaged}`;
  if (cache[cacheKey] !== undefined) return cache[cacheKey];

  let res = 0;
  const char = row.springs[iSprings];
  if (char === '.' || char === '?') {
    if (iContiguousDamaged === 0) {
      res += countArrangementsRecursive(row, iSprings + 1, iGroup, iContiguousDamaged, cache);
    } else if (iContiguousDamaged === row.group[iGroup]) {
      res += countArrangementsRecursive(row, iSprings + 1, iGroup + 1, 0, cache);
    }
  }
  if (char === '#' || char === '?') {
    if (iGroup < row.group.length && iContiguousDamaged < row.group[iGroup]) {
      res += countArrangementsRecursive(row, iSprings + 1, iGroup, iContiguousDamaged + 1, cache);
    }
  }

  cache[cacheKey] = res;
  return res;
}

function countArrangements(row: Row): number {
  return countArrangementsRecursive(row, 0, 0, 0, {});
}

function unfoldRow(row: Row, unfoldingFactor: number): Row {
  const newRow: Row = { springs: row.springs, group: row.group };
  for (let i = 1; i < unfoldingFactor; i++) {
    newRow.springs += '?' + row.springs;
    newRow.group = newRow.group.concat(row.group);
  }
  return newRow;
}

function solve(input: string[]): number {
  const rows = parseInput(input);
  let res = 0;
  for (const row of rows) {
    res += countArrangements(row);
  }
  return res;
}

function readFile(fileName: string): string[] {
  const file = fs.readFileSync(fileName, 'utf8');
  return file.trim().split('\n');
}

const input = readFile('input.txt');
console.log(solve(input));

File: swift/day12_part1_2023.swift

import Foundation

struct Row {
    var springs: String
    var group: [Int]
}

func parseInput(input: [String]) -> [Row] {
    var rows: [Row] = []
    for line in input {
        let parts = line.components(separatedBy: " ")
        let springs = parts[0]
        let ints = parseStringToInts(numbersLine: parts[1])
        
        let row = Row(springs: springs, group: ints)
        rows.append(row)
    }
    return rows
}

func parseStringToInts(numbersLine: String) -> [Int] {
    var numbers: [Int] = []
    let numbersParts = numbersLine.components(separatedBy: ",")
    for numberStr in numbersParts {
        if let number = Int(numberStr) {
            numbers.append(number)
        } else {
            fatalError("Error converting string to int")
        }
    }
    return numbers
}

func countArrangementsRecursive(row: Row, iSprings: Int, iGroup: Int, iContiguousDamaged: Int, cache: inout [Int: Int]) -> Int {
    if iSprings == row.springs.count {
        if iGroup == row.group.count && iContiguousDamaged == 0 {
            return 1
        } else if iGroup == row.group.count - 1 && iContiguousDamaged == row.group[iGroup] {
            return 1
        }
        return 0
    }
    
    let cacheKey = iSprings * 1000 + iGroup * 100 + iContiguousDamaged
    if let val = cache[cacheKey] {
        return val
    }
    
    var res = 0
    let char = row.springs[row.springs.index(row.springs.startIndex, offsetBy: iSprings)]
    if char == "." || char == "?" {
        if iContiguousDamaged == 0 {
            res += countArrangementsRecursive(row: row, iSprings: iSprings + 1, iGroup: iGroup, iContiguousDamaged: iContiguousDamaged, cache: &cache)
        } else if iContiguousDamaged == row.group[iGroup] {
            res += countArrangementsRecursive(row: row, iSprings: iSprings + 1, iGroup: iGroup + 1, iContiguousDamaged: 0, cache: &cache)
        }
    }
    if char == "#" || char == "?" {
        if iGroup < row.group.count && iContiguousDamaged < row.group[iGroup] {
            res += countArrangementsRecursive(row: row, iSprings: iSprings + 1, iGroup: iGroup, iContiguousDamaged: iContiguousDamaged + 1, cache: &cache)
        }
    }
    
    cache[cacheKey] = res
    return res
}

func countArrangements(row: Row) -> Int {
    var cache: [Int: Int] = [:]
    return countArrangementsRecursive(row: row, iSprings: 0, iGroup: 0, iContiguousDamaged: 0, cache: &cache)
}

func unfoldRow(row: Row, unfoldingFactor: Int) -> Row {
    var newRow = Row(springs: row.springs, group: row.group)
    
    for _ in 1..<unfoldingFactor {
        newRow.springs += "?" + row.springs
        newRow.group += row.group
    }
    
    return newRow
}

func solve(input: [String]) -> Int {
    let rows = parseInput(input: input)
    
    var res = 0
    for row in rows {
        res += countArrangements(row: row)
    }
    
    return res
}

func readFile(fileName: String) -> [String] {
    do {
        let fileContent = try String(contentsOfFile: fileName)
        return fileContent.components(separatedBy: "\n")
    } catch {
        fatalError("Error reading file")
    }
}

let input = readFile(fileName: "input.txt")
print(solve(input: input))
